### PR TITLE
Motors protocol when receiver gets disconnected

### LINKED ISSUE #165: Kill motors when receiver gets disconnected
### Describe the bug

We need to make sure that the motors are killed in case the receiver seems to be disconnected for more than a few milliseconds. We had tested it a few weeks ago, and the motors kept spinning after unplugging the receiver. 

### Affected File

_No response_

### Affected Line

_No response_

### To Reproduce

Remove the receiver wires

### Minimum reproducible example

Guys, we should simplify bug reporting! 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jamie-54**
## Main Changes

When the pi hasn't read a new receiver value after a certain amount of time it kills the motors.

## Associated Issues

- Closes #165

## Tests

This has been tested on the quadcopter, when the receiver wire is disconnected the motor stop

### CODE REVIEW

**Comment by alphaville** [comment]
@jamie-54 It would be good to test this outside and see what the range of the RC is. Do the motors stop when the receiver gets physically disconnected, or when there is no signal coming form the RC (e.g., if you turn the RC off)? Or both? I wonder whether it would be better for it to send to the motors a low PWM, just below the hovering value, so that it lands a little more smoothly (imagine if you're flying at 100m and the motors turn off).

File: raspberry/bzzz/read_sbus/read_sbus_from_GPIO_receiver.py [hunk 1]
```diff
@@ -42,11 +42,15 @@ def __init__(self,
...
 
-    def get_radio_data(self):
+    def get_radio_data(self,
+                       max_packet_age_in_ms=500):
         """
         Checks if the radio is connected, determines when the last RC data was received
         and reads the 16 channels of data received from the RC
 
+        :param max_packet_age_in_ms: max age a packet can be im ms. Therefore how long the quadcopter
+                                     can fly since it last read the reciver. defualt 500ms
```
Comment (alphaville) [inline raspberry/bzzz/read_sbus/read_sbus_from_GPIO_receiver.py:52]: typo: receiver

**Review by alphaville — APPROVED** [review]

**Comment by jamie-54** [comment]
> @jamie-54 It would be good to test this outside and see what the range of the RC is. Do the motors stop when the receiver gets physically disconnected, or when there is no signal coming form the RC (e.g., if you turn the RC off)? Or both? I wonder whether it would be better for it to send to the motors a low PWM, just below the hovering value, so that it lands a little more smoothly (imagine if you're flying at 100m and the motors turn off).

I'll check today what happens when remote is switched off, it defiantly works for when the receiver gets physically disconnected. it might be worth doing that, I originally was thinking of just killing the motors encase the receiver wire got damaged when testing, but since its inside the case now that probably won't happen. 

Maybe we should try and implement a fail safe for if the ToF sensor can tell the quadcopter is above 1m of the ground it the radio gets disconnected to spin them at a speed just less than hovering?

**Comment by pdavid747** [comment]
> > @jamie-54 It would be good to test this outside and see what the range of the RC is. Do the motors stop when the receiver gets physically disconnected, or when there is no signal coming form the RC (e.g., if you turn the RC off)? Or both? I wonder whether it would be better for it to send to the motors a low PWM, just below the hovering value, so that it lands a little more smoothly (imagine if you're flying at 100m and the motors turn off).
> 
> I'll check today what happens when remote is switched off, it defiantly works for when the receiver gets physically disconnected. it might be worth doing that, I originally was thinking of just killing the motors encase the receiver wire got damaged when testing, but since its inside the case now that probably won't happen.
> 
> Maybe we should try and implement a fail safe for if the ToF sensor can tell the quadcopter is above 1m of the ground it the radio gets disconnected to spin them at a speed just less than hovering?

yea I think that would be a good idea, we wouldn't want it dropping out of the sky. instead decreasing the motor rotations to slightly less of what it would be for hovering so that it lands smoothly. In the future I guess we could implement a GPS failsafe if we get automated flying to work, getting the quad to "return to home" if the receiver gets disconnected.

**Comment by jamie-54** [comment]
I've updated this so if the receiver is disconnected or the radio is turned off when the quadcopter is above 0.6m (default value) the motors will spin slightly less than hovering (estimated to be 650 radio signal approx. 1/4 full speed) and when the quadcopter is below 0.6m they will stop.

**Review by Runway27 — APPROVED** [review]

**Review by pdavid747 — APPROVED** [review]
looks good to me.

**Pull request merged**